pyflakes/checker.py

```diff
@@ -952,21 +968,27 @@ def IMPORT(self, node):
 
     def IMPORTFROM(self, node):
         if node.module == '__future__':
-            if not self.futuresAllowed:
+            # __future__ can only appear in module/doctest scope and
+            # should not have been disabled already in handleNode and
+            # the scope must not have any other type of binding.
+            if (not isinstance(self.scope, ModuleScope) or
+                    not self.scope._futures_allowed or
+                    any(not isinstance(binding, FutureImportation)
+                        for binding in self.scope.values())):
+                self.scope._futures_allowed = False
```

Comment:
I've added `has_only_future_imports` to name one aspect of it.

Except for that, it is the same as the previous logic, excepting that

`not self.futureAllowed`

is now

`not isinstance(self.scope, ModuleScope) or not self.scope._futures_allowed` .

I could name that compound conditional...

```
@property
def futuresAllowed(self):
    return not isinstance(self.scope, ModuleScope) or not self.scope._futures_allowed
```

;-)